src/algorithms/matching.js

```diff
@@ -0,0 +1,1257 @@
+'use strict';
+
+/**
+ * @fileoverview
+ * Maximal and maximum matching algorithms for undirected graphs.
+ */
+
+import {
+/*jshint ignore:start*/
+  Map,
+  Set,
+/*jshint ignore:end*/
+  zipIterator,
+  genRange,
+} from '../_internals';
+
+/**
+ * Find a maximal cardinality matching in the graph.
+ * A matching is a subset of edges in which no node occurs more than once.
+ * The cardinality of a matching is the number of matched edges.
+ *
+ * ### Notes
+ *
+ * The algorithm greedily selects a maximal matching M of the graph G
+ * (i.e. no superset of M exists). It runs in `O(|E|)` time.
+ *
+ * @param {Graph} G
+ *   Undirected graph
+ * @return {Set} matching
+ *   A maximal matching of the graph.
+ */
+export function maximalMatching(G) {
+  const matching = new Set();
+  const vertices = new Set();
+  for (let [u, v] of G.edgesIter()) {
+    // If the edge isn't covered, add it to the matching
+    // then remove neighborhood of u and v from consideration.
+    if (!vertices.has(u) && !vertices.has(v)) {
+      matching.add([u, v]);
+      vertices.update([u, v]);
```

Comment:
Ok, the original implementation was already `O(|E|)` but this change should divide the asymptotic running time of the method by a constant factor.